File: sdk/python/fleetglass/adapters.py
"""Provider adapters: wrap(client) returns the same client with its completion
method intercepted to auto-capture spans onto the tracer."""
import weakref

# Idempotency: wrap() monkey-patches the client in place, so re-wrapping would
# patch the patched method and double-emit. Track wrapped clients and no-op on re-wrap.
_wrapped = weakref.WeakSet()

def _contents_to_text(contents):
    if isinstance(contents, str):
        return contents
    if isinstance(contents, (list, tuple)):
        out = []
        for c in contents:
            if isinstance(c, str):
                out.append(c)
            else:
                parts = getattr(c, "parts", None) or (c.get("parts") if isinstance(c, dict) else None) or []
                out.append("".join(getattr(p, "text", None) or (p.get("text") if isinstance(p, dict) else "") or "" for p in parts))
        return "\n".join(out)
    return ""

def _sys_text(config):
    if not config:
        return ""
    si = config.get("system_instruction") if isinstance(config, dict) else getattr(config, "system_instruction", None)
    if not si:
        return ""
    if isinstance(si, str):
        return si
    parts = getattr(si, "parts", None) or (si.get("parts") if isinstance(si, dict) else None) or []
    text = "".join((getattr(p, "text", None) or (p.get("text") if isinstance(p, dict) else "") or "") for p in parts)
    return text or str(si)

def _tools_text(config):
    if not config:
        return ""
    tools = config.get("tools") if isinstance(config, dict) else getattr(config, "tools", None)
    return "" if not tools else str(tools)

def _google_messages(contents):
    if isinstance(contents, str):
        return [{"role": "user", "content": contents}]
    if not isinstance(contents, (list, tuple)):
        return []
    out = []
    for c in contents:
        if isinstance(c, str):
            out.append({"role": "user", "content": c})
        else:
            parts = getattr(c, "parts", None) or (c.get("parts") if isinstance(c, dict) else None) or []
            role = getattr(c, "role", None) or (c.get("role") if isinstance(c, dict) else None)
            text = "".join((getattr(p, "text", None) or (p.get("text") if isinstance(p, dict) else "") or "") for p in parts)
            out.append({"role": "assistant" if role == "model" else "user", "content": text})
    return out

def _plain_messages(messages):
    out = []
    for m in messages or []:
        if m.get("role") == "system":
            continue
        c = m.get("content")
        out.append({"role": "assistant" if m.get("role") == "assistant" else "user",
                    "content": c if isinstance(c, str) else ""})
    return out

def _safe_emit(tracer, **fields):
    # Telemetry must never break the agent: a failed emit drops the span, never the call.
    try:
        tracer.emit_chat(**fields)
    except Exception:
        pass

def _wrap_google(client, tracer):
    real = client.models.generate_content

    def traced(model=None, contents=None, config=None, **kw):
        res = real(model=model, contents=contents, config=config, **kw)
        um = getattr(res, "usage_metadata", None)
        history = _contents_to_text(contents)
        _safe_emit(
            tracer,
            model=getattr(res, "model_version", None) or model or "unknown",
            input_tokens=getattr(um, "prompt_token_count", 0) or 0,
            output_tokens=(getattr(um, "candidates_token_count", 0) or 0) + (getattr(um, "thoughts_token_count", 0) or 0),
            prompt=history,
            completion=getattr(res, "text", "") or "",
            context={"system": _sys_text(config), "history": history, "tools": _tools_text(config)},
            request={"system": _sys_text(config), "messages": _google_messages(contents), "tools": (config or {}).get("tools") if isinstance(config, dict) else getattr(config, "tools", None)},
        )
        return res

    client.models.generate_content = traced  # monkey-patch the bound method on this instance
    return client

def _msgs_text(messages):
    for m in reversed(messages or []):
        if m.get("role") == "user":
            c = m.get("content")
            if isinstance(c, str):
                return c
            return "\n".join(b.get("text", "") for b in (c or []) if isinstance(b, dict) and b.get("type") == "text")
    return ""

def _sys_from_messages(messages):
    return "\n".join(m.get("content", "") for m in (messages or []) if m.get("role") == "system" and isinstance(m.get("content"), str))

def _wrap_anthropic(client, tracer):
    real = client.messages.create
    def traced(model=None, system=None, messages=None, **kw):
        res = real(model=model, system=system, messages=messages, **kw)
        u = getattr(res, "usage", None)
        _safe_emit(
            tracer,
            model=getattr(res, "model", None) or model or "unknown",
            input_tokens=(getattr(u, "input_tokens", 0) or 0) + (getattr(u, "cache_read_input_tokens", 0) or 0) + (getattr(u, "cache_creation_input_tokens", 0) or 0),
            output_tokens=getattr(u, "output_tokens", 0) or 0,
            prompt=_msgs_text(messages),
            completion="".join(getattr(b, "text", "") for b in getattr(res, "content", []) if getattr(b, "type", "") == "text"),
            context={"system": system or "", "history": _msgs_text(messages), "tools": str(kw.get("tools") or "")},
            request={"system": system or "", "messages": _plain_messages(messages), "tools": kw.get("tools")},
        )
        return res
    client.messages.create = traced
    return client

def _wrap_openai(client, tracer):
    real = client.chat.completions.create
    def traced(model=None, messages=None, **kw):
        res = real(model=model, messages=messages, **kw)
        u = getattr(res, "usage", None)
        _safe_emit(
            tracer,
            model=getattr(res, "model", None) or model or "unknown",
            input_tokens=getattr(u, "prompt_tokens", 0) or 0,
            output_tokens=getattr(u, "completion_tokens", 0) or 0,
            prompt=_msgs_text(messages),
            completion=(res.choices[0].message.content if getattr(res, "choices", None) else "") or "",
            context={"system": _sys_from_messages(messages), "history": _msgs_text(messages), "tools": str(kw.get("tools") or "")},
            request={"system": _sys_from_messages(messages), "messages": _plain_messages(messages), "tools": kw.get("tools")},
        )
        return res
    client.chat.completions.create = traced
    return client

def wrap(client, tracer):
    if client in _wrapped:
        return client
    models = getattr(client, "models", None)
    if models is not None and hasattr(models, "generate_content"):
        wrapped = _wrap_google(client, tracer)
    else:
        messages = getattr(client, "messages", None)
        chat = getattr(client, "chat", None)
        if messages is not None and hasattr(messages, "create"):
            wrapped = _wrap_anthropic(client, tracer)
        elif chat is not None and hasattr(getattr(chat, "completions", None), "create"):
            wrapped = _wrap_openai(client, tracer)
        else:
            raise TypeError("fleetglass: unrecognized client (google-genai / anthropic / openai)")
    _wrapped.add(wrapped)
    return wrapped

File: sdk/node/adapters.js
// Provider adapters: wrap(client) returns a proxy that auto-captures each
// model call onto fg.emitChat, reading the current agent/parent from context.

// Telemetry must never break the agent: a failed emit drops the span, never the call.
function safeEmit(fg, fields) {
  try { fg.emitChat(fields); } catch { /* drop the span, keep the response */ }
}

function contentsToText(contents) {
  if (typeof contents === 'string') return contents;
  if (!Array.isArray(contents)) return '';
  return contents.map((c) => typeof c === 'string' ? c : (c.parts || []).map((p) => p.text || '').join('')).join('\n');
}

// Provider request → canonical { system, messages:[{role,content}], tools }.
// ponytail: canonical capture is text-only — multimodal and tool-result content blocks are dropped; full-fidelity capture is a later milestone.
function googleMessages(contents) {
  if (typeof contents === 'string') return [{ role: 'user', content: contents }];
  if (!Array.isArray(contents)) return [];
  return contents.map((c) => ({
    role: c.role === 'model' ? 'assistant' : 'user',
    content: typeof c === 'string' ? c : (c.parts || []).map((p) => p.text || '').join(''),
  }));
}
function anthropicMessages(messages) {
  return (messages || []).map((m) => ({
    role: m.role === 'assistant' ? 'assistant' : 'user',
    content: typeof m.content === 'string' ? m.content : (m.content || []).filter((b) => b.type === 'text').map((b) => b.text).join('\n'),
  }));
}
function openaiMessages(messages) {
  return (messages || []).filter((m) => m.role !== 'system').map((m) => ({
    role: m.role === 'assistant' ? 'assistant' : 'user',
    content: typeof m.content === 'string' ? m.content : '',
  }));
}

function wrapGoogle(client, fg) {
  const models = client.models;
  const orig = models.generateContent.bind(models);
  const proxiedModels = new Proxy(models, {
    get(t, p) {
      if (p !== 'generateContent') return t[p];
      return async (req) => {
        const res = await orig(req);
        const um = res.usageMetadata || {};
        const historyText = contentsToText(req.contents);
        safeEmit(fg, {
          model: res.modelVersion || req.model,
          inputTokens: um.promptTokenCount || 0,
          outputTokens: (um.candidatesTokenCount || 0) + (um.thoughtsTokenCount || 0),
          prompt: historyText,
          completion: res.text || '',
          context: {
            system: sysText(req.config?.systemInstruction),
            history: historyText,
            tools: req.config?.tools ? JSON.stringify(req.config.tools) : '',
          },
          request: { system: sysText(req.config?.systemInstruction), messages: googleMessages(req.contents), tools: req.config?.tools },
        });
        return res;
      };
    },
  });
  return new Proxy(client, { get(t, p) { return p === 'models' ? proxiedModels : t[p]; } });
}

function sysText(si) {
  if (!si) return '';
  if (typeof si === 'string') return si;
  return (si.parts || []).map((p) => p.text || '').join('') || String(si);
}

function msgsText(messages) {
  const last = [...(messages || [])].reverse().find((m) => m.role === 'user');
  const c = last?.content;
  return typeof c === 'string' ? c : (c || []).filter((b) => b.type === 'text').map((b) => b.text).join('\n');
}
function sysFromMessages(messages) {
  return (messages || []).filter((m) => m.role === 'system').map((m) => (typeof m.content === 'string' ? m.content : '')).join('\n');
}

function wrapAnthropic(client, fg) {
  const orig = client.messages.create.bind(client.messages);
  const messages = new Proxy(client.messages, { get(t, p) {
    if (p !== 'create') return t[p];
    return async (params) => {
      const res = await orig(params);
      const u = res.usage || {};
      safeEmit(fg, {
        model: res.model || params.model,
        inputTokens: (u.input_tokens || 0) + (u.cache_read_input_tokens || 0) + (u.cache_creation_input_tokens || 0),
        outputTokens: u.output_tokens || 0,
        prompt: msgsText(params.messages),
        completion: (res.content || []).filter((b) => b.type === 'text').map((b) => b.text).join('\n'),
        context: { system: params.system || '', history: msgsText(params.messages), tools: params.tools ? JSON.stringify(params.tools) : '' },
        request: { system: params.system || '', messages: anthropicMessages(params.messages), tools: params.tools },
      });
      return res;
    };
  } });
  return new Proxy(client, { get(t, p) { return p === 'messages' ? messages : t[p]; } });
}

function wrapOpenAI(client, fg) {
  const orig = client.chat.completions.create.bind(client.chat.completions);
  const completions = new Proxy(client.chat.completions, { get(t, p) {
    if (p !== 'create') return t[p];
    return async (params) => {
      const res = await orig(params);
      const u = res.usage || {};
      safeEmit(fg, {
        model: res.model || params.model,
        inputTokens: u.prompt_tokens || 0,
        outputTokens: u.completion_tokens || 0,
        prompt: msgsText(params.messages),
        completion: res.choices?.[0]?.message?.content || '',
        context: { system: sysFromMessages(params.messages), history: msgsText(params.messages), tools: params.tools ? JSON.stringify(params.tools) : '' },
        request: { system: sysFromMessages(params.messages), messages: openaiMessages(params.messages), tools: params.tools },
      });
      return res;
    };
  } });
  const chat = new Proxy(client.chat, { get(t, p) { return p === 'completions' ? completions : t[p]; } });
  return new Proxy(client, { get(t, p) { return p === 'chat' ? chat : t[p]; } });
}

// Idempotency: re-wrapping an already-wrapped client would stack emits.
// Mark the proxies we return and no-op on re-wrap.
const wrapped = new WeakSet();

export function wrap(client, fg) {
  if (wrapped.has(client)) return client;
  let proxy;
  if (client?.models?.generateContent) proxy = wrapGoogle(client, fg);
  else if (client?.messages?.create) proxy = wrapAnthropic(client, fg);
  else if (client?.chat?.completions?.create) proxy = wrapOpenAI(client, fg);
  else throw new Error('fleetglass: unrecognized client (google-genai / anthropic / openai)');
  wrapped.add(proxy);
  return proxy;
}
